### PR TITLE
Display custom data name instead of indices in TileData inspector

### LINKED ISSUE #76733: Custom data layers don't have the correct name displayed in the TileSet window
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Windows 11

### Issue description

Currently in Godot 4 you can create powerful Custom Data Layers. But when applying changes to that data in the tileset window (for example selecting true/false value for specific tiles) the name for specific custom data layer is the index number instead of actual name set in Custom Data Layers.

obraz
obraz

### Steps to reproduce

1. Create tilemap node
2. Create/assign tileset to that tilemap node
3. Create new Custom Data Layer for that TileSet
4. Go to TileSet window, press select and select any tile in the tileset
5. See that custom data layer works, but has incorrect name

### Minimal reproduction project

N/A

### LINKED ISSUE #92149: Tileset custom data shows as index (not name) in select mode
### Tested versions

- Reproducible in v4.3.dev6

### System information

Godot v4.3.dev6 - Windows 10.0.19045 - GLES3 (Compatibility) - NVIDIA GeForce RTX 3070 (NVIDIA; 31.0.15.3619) - 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz (16 Threads)

### Issue description

In the tileset menu, when setting custom data from the select tab, custom data shows as an index, rather than the name of that data element:
image

 Custom data is correctly shown by name when in the paint tab:
image

I believe this is a regression from 4.2, where custom data layers were always shown by name. I'd imagine that with more custom data layers it may be difficult to tell which is which from the select tab.

### Steps to reproduce

- Create a tileset
- Give the tileset a custom data layer, name it something meaningful
- Open the select tab and unfold the custom data dropdown
- The label for the data shows the index rather than the name you set for the custom data
- In the paint tab, the data is correctly labelled with your custom data name

### Minimal reproduction project (MRP)

N/A

### PR DESCRIPTION
**Pull request by groud**
Supersedes https://github.com/godotengine/godot/pull/66853
Closes https://github.com/godotengine/godot/issues/92149
Closes https://github.com/godotengine/godot/issues/76733

### CODE REVIEW

File: doc/classes/EditorInspectorPlugin.xml [hunk 1]
```diff
@@ -77,7 +77,8 @@
 			<return type="void" />
 			<param index="0" name="property" type="String" />
 			<param index="1" name="editor" type="Control" />
-			<param index="2" name="add_to_end" type="bool" default="false" />
+			<param index="2" name="add_to_end" type="bool" />
+			<param index="3" name="label" type="String" default="false" />
```
Comment (akien-mga) [inline doc/classes/EditorInspectorPlugin.xml:81]:
```suggestion
			<param index="2" name="add_to_end" type="bool" default="false" />
			<param index="3" name="label" type="String" default="&quot;&quot;" />
```

You'll also need to register a compat method binding for GDExtension.

File: editor/editor_inspector.compat.inc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**************************************************************************/
+
+#ifndef DISABLE_DEPRECATED
+
+void EditorInspectorPlugin::_add_property_editor_92322(const String &p_for_property, Control *p_prop, bool p_add_to_end) {
+	AddedEditor ae;
+	ae.properties.push_back(p_for_property);
+	ae.property_editor = p_prop;
+	ae.add_to_end = p_add_to_end;
+	added_editors.push_back(ae);
```
Comment (akien-mga) [inline editor/editor_inspector.compat.inc:38]:
```suggestion
	add_property_editor(p_for_property, p_prop, p_add_to_end, "");
```

File: editor/editor_inspector.compat.inc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+void EditorInspectorPlugin::_bind_compatibility_methods() {
+	ClassDB::bind_compatibility_method(D_METHOD("add_property_editor", "property", "editor", "add_to_end"), &EditorInspectorPlugin::_add_property_editor_92322, DEFVAL(false));
```
Comment (akien-mga) [inline editor/editor_inspector.compat.inc:42]: The standard naming scheme is:
```suggestion
	ClassDB::bind_compatibility_method(D_METHOD("add_property_editor", "property", "editor", "add_to_end"), &EditorInspectorPlugin::_add_property_editor_bind_compat_92322, DEFVAL(false));
```

File: doc/classes/EditorInspectorPlugin.xml [hunk 4]
```diff
@@ -78,6 +78,7 @@
 			<param index="0" name="property" type="String" />
 			<param index="1" name="editor" type="Control" />
 			<param index="2" name="add_to_end" type="bool" default="false" />
+			<param index="3" name="label" type="String" default="&quot;&quot;" />
```
Comment (KoBeWi) [inline doc/classes/EditorInspectorPlugin.xml:81]: The new argument could be mentioned in the description.

File: editor/plugins/tiles/tile_set_atlas_source_editor.cpp [hunk 5]
```diff
@@ -2938,6 +2938,18 @@ bool EditorInspectorPluginTileData::parse_property(Object *p_object, const Varia
...
 			return true;
 		}
+	} else if (p_path.begins_with("custom_data_") && p_path.trim_prefix("custom_data_").is_valid_int()) {
+		// Custom data layers.
+		int layer_index = components[0].trim_prefix("custom_data_").to_int();
+		ERR_FAIL_COND_V(layer_index < 0, false);
+		EditorProperty *ep = EditorInspectorDefaultPlugin::get_editor_for_property(p_object, p_type, p_path, PROPERTY_HINT_NONE, "", PROPERTY_USAGE_DEFAULT);
+		const TileSetAtlasSourceEditor::AtlasTileProxyObject *proxy_obj = Object::cast_to<TileSetAtlasSourceEditor::AtlasTileProxyObject>(p_object);
+		const TileSetAtlasSource *atlas_source = *proxy_obj->get_edited_tile_set_atlas_source();
+		ERR_FAIL_COND_V(!atlas_source, false);
```
Comment (KoBeWi) [inline editor/plugins/tiles/tile_set_atlas_source_editor.cpp:2948]:
```suggestion
		ERR_FAIL_NULL_V(atlas_source, false);
```

File: editor/plugins/tiles/tile_set_atlas_source_editor.cpp [hunk 6]
```diff
@@ -2938,6 +2938,18 @@ bool EditorInspectorPluginTileData::parse_property(Object *p_object, const Varia
...
+		const TileSet *tile_set = atlas_source->get_tile_set();
+		ERR_FAIL_COND_V(!tile_set, false);
```
Comment (KoBeWi) [inline editor/plugins/tiles/tile_set_atlas_source_editor.cpp:2950]:
```suggestion
		ERR_FAIL_NULL(tile_set, false);
```

**Review by KoBeWi — APPROVED** [review]

File: misc/extension_api_validation/4.2-stable.expected [hunk 7]
```diff
@@ -357,3 +357,8 @@ Validate extension JSON: Error: Field 'classes/RenderingServer/methods/canvas_it
 Validate extension JSON: Error: Field 'classes/RenderingServer/methods/canvas_item_add_rect/arguments': size changed value in new API, from 3 to 4.
 
 Optional arguments added. Compatibility methods registered.
+
+
+GH-92322
+--------
+Validate extension JSON: Error: Field 'classes/EditorInspectorPlugin/methods/add_property_editor/arguments': size changed value in new API, from 3 to 4.
```
Comment (AThousandShips) [inline misc/extension_api_validation/4.2-stable.expected:364]: Please add some note about what was done, and that compatibility a method was added

**Review by AThousandShips — APPROVED** [review]

**Pull request merged**